tests/components/permobil/test_config_flow.py

```diff
@@ -0,0 +1,298 @@
+
+import pytest
+
+from homeassistant import config_entries
+from homeassistant.components.permobil import config_flow
+from homeassistant.components.permobil.const import DOMAIN
+from homeassistant.const import CONF_CODE, CONF_EMAIL, CONF_REGION, CONF_TOKEN, CONF_TTL
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+MOCK_URL = "https://example.com"
+MOCK_REGION = "region_name"
+MOCK_REGIONS = {MOCK_REGION: MOCK_URL}
+MOCK_TOKEN = ("a" * 256, "date")
+MOCK_CODE = "012345"
+MOCK_EMAIL = "valid@email.com"
+INVALID_EMAIL = "this is not a valid email"
+INVALID_CODE = "this is not a valid code"
+MOCK_REENTRY = MagicMock()
+MOCK_REENTRY.data = {CONF_EMAIL: MOCK_EMAIL, CONF_REGION: MOCK_URL}
+MOCK_CONTEXT = {"entry_id": None}
+
+
+class MockApiClientError:
+    """An instance of the api that always returns an client exception when called."""
+
+    def set_email(self):
+        """Raise exception."""
+        raise config_flow.MyPermobilClientException
+
+    def set_code(self):
+        """Raise exception."""
+        raise config_flow.MyPermobilClientException
+
+
+class MockApiAPIError:
+    """An instance of the api that always returns an API exception when called."""
+
+    region = ""
+
+    def set_region(self):
+        """Pass."""
+
+    def set_code(self):
+        """Pass."""
+
+    def set_email(self):
+        """Pass."""
+
+    async def request_application_code():
+        """Raise MyPermobilAPIException."""
+        raise config_flow.MyPermobilAPIException
+
+    async def request_region_names():
+        """Raise MyPermobilAPIException."""
+        raise config_flow.MyPermobilAPIException
+
+    async def request_application_token():
+        """Raise MyPermobilAPIException."""
+        raise config_flow.MyPermobilAPIException
+
+
+class MockApiSuccess:
+    """An instance of the api that never complains."""
+
+    region = ""
+
+    def set_region(self):
+        """Pass."""
+
+    def set_code(self):
+        """Pass."""
+
+    def set_email(self):
+        """Pass."""
+
+    async def request_region_names():
+        """Fake a request for regions."""
+        return MOCK_REGIONS
+
+    async def request_application_code():
+        """Fake a request for application code."""
+
+    async def request_application_token():
+        """Return a mock token."""
+        return MOCK_TOKEN
+
+
+async def test_flow_init(hass: HomeAssistant) -> None:
+    """Test config flow init."""
+    with patch(
+        "homeassistant.components.permobil.config_flow.MyPermobil",
+        lambda x, session: MockApiSuccess,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+
+async def test_form_invalid_region_api(hass: HomeAssistant) -> None:
```

Comment:
Please create tests like https://github.com/home-assistant/core/blob/06f97679ee8b7f675ae53dbd9a7990e1a2993501/tests/components/unifi/test_config_flow.py#L96

Please don't mock PermobilConfigFlow variables, which are filled from previous steps.

Reply:
I overhauled the tests to closer mimic a whole config flow. I hope that this addresses the feedback you gave.